compiler/javac/src/org/jetbrains/kotlin/javac/Javac.kt

```diff
@@ -0,0 +1,227 @@
+import java.io.Closeable
+import java.io.File
+import javax.lang.model.element.TypeElement
+import javax.tools.JavaFileManager
+import javax.tools.JavaFileObject
+import javax.tools.StandardLocation
+
+class Javac(javaFiles: Collection<File>,
+            classPathRoots: List<File>,
+            outDir: File?,
+            private val messageCollector: MessageCollector?,
+            arguments: Array<String>?) : Closeable {
+
+    companion object {
+        fun getInstance(project: Project): Javac = ServiceManager.getService(project, Javac::class.java)
+    }
+
+    val JAVA_LANG_OBJECT by lazy { findClassInSymbols(CommonClassNames.JAVA_LANG_OBJECT) }
+
+    private val context = Context()
+
+    init {
+        messageCollector?.let { JavacLogger.preRegister(context, it) }
+        arguments?.toList()?.let { JavacOptionsMapper.map(Options.instance(context), it) }
+    }
+
+    private val javac = object : JavaCompiler(context) {
+        override fun parseFiles(files: Iterable<JavaFileObject>?) = compilationUnits
+    }
+    private val fileManager = context[JavaFileManager::class.java] as JavacFileManager
+
+    init {
+        fileManager.setLocation(StandardLocation.CLASS_PATH, classPathRoots)
+        outDir?.let {
+            it.mkdirs()
+            fileManager.setLocation(StandardLocation.CLASS_OUTPUT, listOf(it))
+        }
+    }
+
+    private val symbols = Symtab.instance(context)
+    private val trees = JavacTrees.instance(context)
+    private val elements = JavacElements.instance(context)
+    private val fileObjects = fileManager.getJavaFileObjectsFromFiles(javaFiles).toJavacList()
+    private val compilationUnits: JavacList<JCTree.JCCompilationUnit> = fileObjects.map(javac::parse).toJavacList()
+
+    private val javaClasses = compilationUnits
+            .flatMap { unit -> unit.typeDecls
+                    .flatMap { JCClass(it as JCTree.JCClassDecl, trees.getPath(unit, it), this).withInnerClasses() }
+            }
+            .map { it.fqName to it }
+            .toMap()
+
+    private val javaPackages = compilationUnits
+            .map { FqName(it.packageName.toString()) to JCPackage(it.packageName.toString(), this) }
+            .toMap()
+
+    fun compile() = with(javac) {
+        if (errorCount() > 0) return false
+
+        fileManager.setClassPathForCompilation()
+        messageCollector?.report(CompilerMessageSeverity.INFO,
+                                 "Compiling Java sources",
+                                 CompilerMessageLocation.NO_LOCATION)
+        compile(fileObjects)
+        errorCount() == 0
+    }
+
+    override fun close() {
+        fileManager.close()
+        javac.close()
+    }
+
+    fun findClass(fqName: FqName, scope: GlobalSearchScope = EverythingGlobalScope()) = when {
+        scope is EverythingGlobalScope -> javaClasses[fqName] ?: findClassInSymbols(fqName.asString())
+        scope.contains(AnyJavaSourceVirtualFile) -> javaClasses[fqName]
```

Comment:
In this case `contains` function just checks a file extension. This is why we can use this mock object here.